PolyEngine/Engine/Src/ECS/World.cpp

```diff
@@ -7,42 +7,49 @@ World::World()
 	: EntitiesAllocator(MAX_ENTITY_COUNT)
 {
 	memset(ComponentAllocators, 0, sizeof(IterablePoolAllocatorBase*) * MAX_COMPONENTS_COUNT);
-	memset(WorldComponents, 0, sizeof(ComponentBase*) * MAX_WORLD_COMPONENTS_COUNT);
 
-	rootEntity = SpawnEntityInternal();
+	RootEntity = SpawnEntityInternal();
 }
 
 //------------------------------------------------------------------------------
 World::~World()
 {
-	// copy entities
-	if(rootEntity)
-		DestroyEntity(rootEntity.Get());
+	if (RootEntity)
+	{
+		// destroy entities
+		auto childrenCopy = RootEntity->GetChildren();
+		for (Entity* child : childrenCopy)
+			DestroyEntity(child);
+
+		// remove world components
+		for (size_t i = 0; i < MAX_WORLD_COMPONENTS_COUNT; i++)
+			if (RootEntity->Components[i])
+				delete (RootEntity->Components[i]);
+
+		// destroy root
+		RootEntity->~Entity();
+		delete RootEntity.Get();
```

Comment:
If we unify WorldComponents and Components (as we should) this change (as in this block of code) is entirely unnecessary, as destruction of world components would be handled by normal ComponentAllocators deletion.